### PR TITLE
Manually run GC on reactor tick.

### PR DESCRIPTION
**Pull request by erikjohnston**
This also adds a metric for amount of time spent in GC.

### CODE REVIEW

File: synapse/metrics/__init__.py [hunk 1]
```diff
@@ -182,6 +184,18 @@ def f(*args, **kwargs):
...
+        threshold = gc.get_threshold()
+        counts = gc.get_count()
+
+        start = time.time() * 1000
+        for i in [2, 1, 0]:
+            if threshold[i] < counts[i]:
+                logger.info("Collecting gc %d", i)
+                gc.collect(i)
+        end = time.time() * 1000
+        gc_time.inc_by(end - start)
```
Comment (NegativeMjark) [inline synapse/metrics/__init__.py]: Might be interesting to collect stats about the different levels of collection going on.

Comment (erikjohnston) [inline synapse/metrics/__init__.py]: Oooh, yes, good point.

**Comment by ara4n** [comment]
this seems really weird - surely GCs can be very costly with lots of objects, especially for the higher generations?  Why would you want to run it on every reactor tick?  Wouldn't it be better to bump up the GC thresholds and just have the machine stop the world to GC once every few hours or whatever?

**Comment by erikjohnston** [comment]
@ara4n We check if we need to run the GC on each reactor tick, we don't necessarily do so. The check should be equivalent to the one done internally by python

**Comment by ara4n** [comment]
okay... but what is the rationale for doing so? (aka where are the comments?)

**Comment by erikjohnston** [comment]
Sure I can add a comment. It's mainly so that we can get some metrics over how long we spend GC'ing

**Comment by ara4n** [comment]
aaah! makes much more sense - thanks :)

**Comment by NegativeMjark** [comment]
LGTM. Might want to be careful about merging this to make sure it doesn't end up on one of our servers by accident.

**Pull request merged**